src/messages/embed.ts

```diff
@@ -0,0 +1,300 @@
+import type {
+	APIEmbed,
+	APIEmbedAuthor,
+	APIEmbedField,
+	APIEmbedFooter,
+	APIEmbedImage,
+	APIEmbedProvider,
+	APIEmbedThumbnail,
+	APIEmbedVideo,
+} from 'discord-api-types/v9';
+import ow from 'ow';
+
+const embedFieldPredicate = ow.object.exactShape({
+	name: ow.string.minLength(1).maxLength(256),
+	value: ow.string.minLength(1).maxLength(1024),
+	inline: ow.optional.boolean,
+});
+
+const embedFieldsArrayPredicate = ow.array.ofType(embedFieldPredicate);
+
+function validateFieldLength(fields: APIEmbedField[], amountAdding: number): void {
+	ow(fields.length + amountAdding, 'field amount', ow.number.lessThanOrEqual(25));
+}
+
+/**
+ * Represents an embed in a message (image/video preview, rich embed, etc.)
+ */
+export class Embed implements APIEmbed {
+	/**
+	 * An array of fields of this embed.
+	 */
+	public fields: APIEmbedField[];
+
+	/**
+	 * The embed title.
+	 */
+	public title?: string;
+
+	/**
+	 * The embed description.
+	 */
+	public description?: string;
+
+	/**
+	 * The embed url.
+	 */
+	public url?: string;
+
+	/**
+	 * The embed color.
+	 */
+	public color?: number;
+
+	/**
+	 * The timestamp of the embed in the ISO format.
+	 */
+	public timestamp?: string;
+
+	/**
+	 * The embed thumbnail data.
+	 */
+	public thumbnail?: APIEmbedThumbnail;
+
+	/**
+	 * The embed image data.
+	 */
+	public image?: APIEmbedImage;
+
+	/**
+	 * Received video data.
+	 */
+	public video?: APIEmbedVideo;
+
+	/**
+	 * The embed author data.
+	 */
+	public author?: APIEmbedAuthor;
+
+	/**
+	 * Received data about the embed provider.
+	 */
+	public provider?: APIEmbedProvider;
+
+	/**
+	 * The embed footer data.
+	 */
+	public footer?: APIEmbedFooter;
+
+	public constructor(data: APIEmbed = {}) {
+		this.title = data.title;
+		this.description = data.description;
+		this.url = data.url;
+		this.color = data.color;
+		this.thumbnail = data.thumbnail;
+		this.image = data.image;
+		this.video = data.video;
+		this.author = data.author;
+		this.provider = data.provider;
+		this.footer = data.footer;
+		this.fields = data.fields ?? [];
+
+		if (data.timestamp) this.timestamp = new Date(data.timestamp).toISOString();
+	}
+
+	/**
+	 * The accumulated length for the embed title, description, fields, footer text, and author name.
+	 */
+	public get length(): number {
+		return (
+			(this.title?.length ?? 0) +
+			(this.description?.length ?? 0) +
+			(this.fields.length >= 1
+				? this.fields.reduce((prev, curr) => prev + curr.name.length + curr.value.length, 0)
+				: 0) +
+			(this.footer?.text.length ?? 0) +
+			(this.author?.name?.length ?? 0)
+		);
+	}
+
+	/**
+	 * Adds a field to the embed (max 25).
+	 * @param name The name of this field.
+	 * @param value The value of this field.
+	 * @param inline If this field will be displayed inline.
+	 */
+	public addField(name: string, value: string, inline = false): this {
+		// Ensure adding this field won't exceed the 25 field limit
+		validateFieldLength(this.fields, 1);
```

Comment:
This is redundant (line 142 is already validating this), this also makes the behavior of `addField` and `addFields` differ if there already are 25 fields and a 26th, but invalid, field is added.
The former throws an error mentioning the length of the field array, the latter throws an error mentioning that the field is not valid.